File: Leetcode/542. 01 Matrix.ts
function updateMatrix(mat: number[][]): number[][] {
    let f: boolean[][] = [];
    let ans: number[][] = [];
    let q: number[][] = [];
    
    for(let i=0; i<mat.length; i++) {
        f.push([]);
        ans.push([])
        for(let j=0; j<mat[i].length; j++) {
            f[i].push(false);
            ans[i].push(0);
            if(mat[i][j] === 0) {
                q.push([i,j]);
                f[i][j] = true;
            }
        }
    }
    
    while(q.length) {
        let [x,y] = q.shift();
        if(x-1>=0 && !f[x-1][y]) {
            ans[x-1][y] = mat[x-1][y]===0 ? 0 : ans[x][y] + 1;
            q.push([x-1,y]);
            f[x-1][y] = true;
        }
        if(y+1<mat[x].length && !f[x][y+1]) {
            ans[x][y+1] = mat[x][y+1]===0 ? 0 : ans[x][y] + 1;
            q.push([x,y+1]);
            f[x][y+1] = true;
        }
        if(x+1<mat.length && !f[x+1][y]) {
            ans[x+1][y] = mat[x+1][y]===0 ? 0 : ans[x][y] + 1;
            q.push([x+1,y]);
            f[x+1][y] = true;
        }
        if(y-1>=0 && !f[x][y-1]) {
            ans[x][y-1] = mat[x][y-1]===0 ? 0 : ans[x][y] + 1;
            q.push([x,y-1]);
            f[x][y-1] = true;
        }
    }
    
    return ans;    
};

File: Leetcode/542. 01 Matrix.cpp
class Solution {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
        int n = mat.size(), m = mat[0].size();
        vector<vector<int>> ans(n, vector<int> (m,-1));
        queue<pair<int,int>> q;
        
        for(int i=0; i<n; i++) {
            for(int j=0; j<m; j++) {
                if(mat[i][j] == 0) {
                    q.push({i,j});
                    ans[i][j] = 0;
                }
            }
        }
        
        while(!q.empty()) {
            pair<int,int> p = q.front();
            q.pop();
            int x = p.first, y = p.second;
            
            if(x-1 >= 0 && ans[x-1][y] == -1) {
                ans[x-1][y] = ans[x][y] + 1;
                q.push({x-1,y});
            }
            if(x+1 < n && ans[x+1][y] == -1) {
                ans[x+1][y] = ans[x][y] + 1;
                q.push({x+1,y});
            }
            if(y-1 >= 0 && ans[x][y-1] == -1) {
                ans[x][y-1] = ans[x][y] + 1;
                q.push({x,y-1});
            }
            if(y+1 < m && ans[x][y+1] == -1) {
                ans[x][y+1] = ans[x][y] + 1;
                q.push({x,y+1});
            }
        }
        
        return ans;
    }
};
